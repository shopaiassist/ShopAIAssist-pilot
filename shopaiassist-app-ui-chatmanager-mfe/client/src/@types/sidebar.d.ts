/**
 * Enumeration of available sorting options for lists.
 */
export type SortOptions = 'by_name' | 'by_date' | 'by_type';

/**
 * Defines the type of items, which can be either a folder or a chat.
 */
export type ItemType = 'folder' | 'chat';

/**
 * Defines the structure of a tree item.
 * @interface
 * @property {string} name - The name of the tree item.
 * @property {string} treeItemId - The unique identifier of the tree item.
 * @property {ItemType} type - The type of the item, distinguishing between chat and folder.
 * @property {Date} createdAt - The creation date of the tree item.
 * @property {Date} updatedAt - The update date of the tree item.
 * @property {string} [parentId] - Optional. The ID of the parent folder, if any.
 */
interface TreeItem {
  name: string;
  treeItemId: string;
  type: ItemType;
  createdAt: Date;
  updatedAt: Date;
  fileCollectionId: string;
  parentId?: string;
}

/**
 * Extends TreeItem to define the structure of a folder item, adding folder-specific properties.
 * @interface
 * @property {string} [matterId] - Optional. A external matterId for the folder.
 * @property {string} [description] - Optional. A description of the folder's contents or purpose.
 * @property {string} [isArchived] - Optional. A boolean for if the folders is archived.
 */
interface FolderItem extends TreeItem {
  matterId?: string;
  description?: string;
  isArchived?: boolean;
}
