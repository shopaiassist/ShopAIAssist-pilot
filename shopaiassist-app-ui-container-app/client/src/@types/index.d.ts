interface AngularMfeLoaderConfig<P extends object = object> extends MfeLoaderConfig<P> {
  /* Specific element id to give the containing div */
  elementId: string;
}

type ReactMfeLoaderConfig<P extends object = object> = MfeLoaderConfig<P>;

interface MfeLoaderConfig<P extends object = object> extends MfeConfig {
  /** The props to pass to the component */
  props?: P;
}

interface MfeConfig {
  /** The name of the module to load from a MFE see 'loadComponent' for more details */
  module: string;
  /** The name of the scope to load from a MFE see 'loadComponent' for more details */
  scope: MfeName;
  /** The url of the MFE to load */
  url: string;
}

type MfeName = 'ai_assistant_mfe' | 'ai_skills_mfe' | 'matters_mfe' | 'file_management_mfe';

type SecondarySideNavTree = 'manage' | 'admin' | 'support' | 'legal';

interface SecondaryNavState {
  show: boolean;
  tree: SecondarySideNavTree;
}

interface NavMenuItem {
  disabled?: boolean;
  icon?: string;
  id: string;
  label: string;
  /** Internal path to navigate to */
  navPath?: string;
  onClick?: () => void;
  secondaryNavTree?: SecondarySideNavTree;
  showArrow?: boolean;
  bold?: boolean;
  /** External URL to link to */
  url?: string;
  pendoId?: string;
}

type NavMenuTree = Record<string, NavMenuItem[]>;

interface BreadcrumbItem {
  label: string;
  href?: string;
}

type TimeoutState = 'expired' | 'warning' | 'active';
