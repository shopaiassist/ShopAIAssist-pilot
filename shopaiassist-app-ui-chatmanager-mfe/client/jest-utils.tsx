import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
