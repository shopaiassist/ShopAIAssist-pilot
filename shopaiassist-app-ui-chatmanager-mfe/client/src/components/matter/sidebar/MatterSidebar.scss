@import "../../../utils/styles/imports";

$headerHeight: 3.75rem;

.matter-sidebar-container {
   display: flex;
   width: 21rem;
   z-index: map-get($z-index-map, "default");

   &.compact-mode {
      position:absolute;
      width: 100%;
   }

   &.full-mode {
      position:relative;
      width: 21rem;
      max-width: 100%;
   }

   .matter-sidebar-nav-header {
      display: flex;
      padding: 0.5rem;
      justify-content: space-between;
      align-items: center;
      align-self: stretch;
      min-height: 3.688rem;
      box-sizing: border-box;
      width: 100%;
      border: rem-calc(1);
      border-bottom: $saf-line-width-thin solid $saf-color-line-subtle;
      border-left: $saf-line-width-thin solid $saf-color-line-subtle;

      .matter-sidebar-nav-back {
         display: flex;
         gap: 0.5rem;
      }
   }

   .matter-sidebar {
      display: flex;
      flex-direction: column;
      align-items: center;
      box-shadow: $saf-drop-shadow-level-3;
      height: calc(100vh - $headerHeight);
      width: 21rem;
      background-color: $saf-color-background-subtle;

      &.compact-mode {
         width: 100%;
      }

      &.full-mode {
         width: 21rem;
         max-width: 100%;
      }

      .matter-sidebar-folder-name {
         padding: 0.5rem;
         margin: 0.75rem;
         display: flex;
         align-self: flex-start;
      }

      .matter-sidebar-header {
         display: flex;
         flex-direction: row;
         justify-content: space-between;
         width: 95%;
         padding: 0.5rem;
      }

      .matter-sidebar-footer {
         display: flex;
         width: 100%;
      }
   }
}