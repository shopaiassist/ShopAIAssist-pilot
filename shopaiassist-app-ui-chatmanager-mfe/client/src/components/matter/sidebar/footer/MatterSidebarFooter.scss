@import "../../../../utils/styles/imports";

.matter-sidebar-footer {
    display: flex;
    flex-direction: column;
    width: 100%;
    box-sizing: border-box;
    padding: 0.3rem;
    

    &-separator {
        border: 0;
        border-bottom: 1px solid $saf-color-gray-200;
        margin: 1rem 0;
        width: 100%;
    }

    .matter-sidebar-archived-matters-selectable {
        width: 100%;

        &-section {
           display: flex;
        }

        &-item {
           display: flex;
           justify-content: space-between;
           align-items: center;
           width: 100%;
        }

        &-icon {
           margin-right: 0.4rem;
        }
     }

    .matter-sidebar-footer-button-container {
        display: flex;
        flex-direction: column;
        align-items: center;
        row-gap: 0.5rem;
        padding: 0.5rem;
        justify-content: center;
    }

    .matter-settings-button {
        width: fit-content;
    }
}