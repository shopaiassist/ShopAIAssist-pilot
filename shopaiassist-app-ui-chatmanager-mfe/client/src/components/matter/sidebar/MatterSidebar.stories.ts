// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
import MatterSidebar from './MatterSidebar';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Matter/Sidebar',
  component: MatterSidebar,
  parameters: {
    layout: 'padded'
  },
  tags: ['autodocs']
} satisfies Meta<typeof MatterSidebar>;

export default meta;

export const Default: StoryObj = { args: {} };
