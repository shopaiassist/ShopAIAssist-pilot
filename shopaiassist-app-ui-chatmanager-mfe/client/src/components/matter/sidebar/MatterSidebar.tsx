import { Text } from '@on/core-components/react';
import MatterSidebarHeader from './header/MatterSidebarHeader';
import ChatList from './chat-list/ChatList';
import { FolderItem, TreeItem } from '../../../@types/sidebar';
import MatterSidebarNavigation from './header/MatterSidebarNavigation';
import useMeasure from 'react-use-measure';
import { animated, useSpring } from '@react-spring/web';
import { useChatSidebarStore } from '../../../store/useChatSidebarStore';
import MatterSidebarFooter from './footer/MatterSidebarFooter';

import './MatterSidebar.scss';
import classNames from 'classnames';
import { ModalType } from '../../../App';
import { useMatters } from '../../../views/Context';

interface MatterSidebarProps {
  isOpen: boolean;
  tree?: (TreeItem | FolderItem)[];
  onNewChat: () => void;
  closeSidebar: () => void;
}

/**
 * A responsive sidebar component for displaying a tree of items such as chats and folders.
 * This sidebar can be toggled to show or hide using an animated transition.
 * The component adapts to changes in the sidebar's width and employs animations for smooth transitions.
 *
 * @param {MatterSidebarProps} props - The properties for the MatterSidebar component.
 * @param {(TreeItem | FolderItem)[]} [props.tree] - Data structure representing the tree items to be displayed in the sidebar.
 * @param {() => void} props.onNewChat - Callback function that triggers when a new chat is initiated.
 * @returns {JSX.Element} The animated sidebar containing a header and a list of chat or folder items.
 *
 * @example
 * <MatterSidebar tree={treeData} onNewChat={handleNewChat} />
 */
const MatterSidebar = ({ isOpen, tree, onNewChat, closeSidebar }: MatterSidebarProps) => {
  const [measureRef, { width }] = useMeasure();
  const springProps = useSpring({
    transform: !isOpen ? `translateX(-${width}px)` : 'translateX(0px)',
    marginRight: !isOpen ? `-${width}px` : '0px'
  });
  const [
    activeFolder,
    setIsMattersettingsActive,
    setIsMatterFilesActive,
    setActiveChat,
    archivedMattersView,
    toggleArchivedMattersView
  ] = useChatSidebarStore((state) => [
    state.activeFolder,
    state.setIsMattersettingsActive,
    state.setIsMatterFilesActive,
    state.setActiveChat,
    state.archivedMattersView,
    state.toggleArchivedMattersView
  ]);
  const { modalType } = useMatters();

  return (
    <animated.div
      style={springProps}
      className={classNames('matter-sidebar-container', {
        'compact-mode': modalType === ModalType.Compact,
        'full-mode': modalType !== ModalType.Compact
      })}
    >
      <div
        ref={measureRef}
        className={classNames('matter-sidebar', {
          'compact-mode': modalType === ModalType.Compact,
          'full-mode': modalType !== ModalType.Compact
        })}
        data-testid="matter-sidebar"
        is-open={isOpen.toString()}
      >
        <MatterSidebarNavigation
          className="matter-sidebar-nav-header"
          isSubsectionView={!!activeFolder || !!archivedMattersView}
          closeSidebar={closeSidebar}
        />
        {activeFolder && (
          <div className="matter-sidebar-folder-name">
            <Text appearance="heading-lg">{activeFolder?.name}</Text>
          </div>
        )}
        {archivedMattersView && (
          <div className="matter-sidebar-folder-name">
            <Text appearance="heading-lg">Archived Matters</Text>
          </div>
        )}
        <MatterSidebarHeader
          className="matter-sidebar-header"
          onNewChat={() => {
            onNewChat();
            if (modalType === ModalType.Compact) {
              closeSidebar();
            }
          }}
          showAddChatButton={!archivedMattersView}
          isFolderView={!!activeFolder}
        />
        <ChatList chatsAndFolders={tree} closeSidebar={closeSidebar} />

        {/* <MatterSidebarFooter
          activeFolder={activeFolder}
          archivedMattersView={archivedMattersView}
          onOpenSettings={() => {
            setIsMattersettingsActive(true), setIsMatterFilesActive(false);
          }}
          onOpenMatterFiles={() => {
            setActiveChat();
            setIsMatterFilesActive(true);
          }}
          toggleArchivedMattersView={toggleArchivedMattersView}
        /> */}
      </div>
    </animated.div>
  );
};

export default MatterSidebar;
