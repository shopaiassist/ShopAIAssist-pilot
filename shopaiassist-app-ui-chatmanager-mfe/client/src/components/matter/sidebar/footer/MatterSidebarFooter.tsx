import { Icon, Text } from '@on/core-components/react';
import Translate from '../../../i18n/Translate';
import { SelectableItem } from '../../../common';
import useMatterFooter from '../../../../hooks/useMatterFooter';
import { FolderItem } from '../../../../@types/sidebar';

import './MatterSidebarFooter.scss';

interface MatterSidebarFooterProps {
  activeFolder?: FolderItem;
  archivedMattersView?: boolean;
  onOpenSettings: () => void;
  onOpenMatterFiles: () => void;
  toggleArchivedMattersView: () => void;
}

/**
 * A footer component for the sidebar. It renders content depending on
 * the position inside the tree and the corresponding available options.
 *
 * @param {MatterSidebarFooterProps} props - The properties for the MatterSidebarFooter component.
 * @param {FolderItem} props.activeFolder - Object with the active folder data.
 * @param {boolean} props.archivedMattersView - Flag to indicate whether a archived matters view is active.
 * @param {() => void} props.onOpenSettings - Callback function that triggers matter settings view to be active.
 * @returns {JSX.Element} The sidebar footer containing options for the current context.
 *
 * @example
 * <MatterSidebarFooter isFolderView={true} onOpenSettings={handleOpenSettings} />
 */
const MatterSidebarFooter = ({
  activeFolder,
  archivedMattersView,
  onOpenSettings,
  onOpenMatterFiles,
  toggleArchivedMattersView
}: MatterSidebarFooterProps) => {
  const [isMatterFilesActive, isMatterSettingsActive] = useMatterFooter();
  const archiveSectionActive = !archivedMattersView && !activeFolder;
  const matterSettingsActive = !!activeFolder;

  return (
    <div className="matter-sidebar-footer">
      {(archiveSectionActive || matterSettingsActive) && <hr className="matter-sidebar-footer-separator" />}

      {archiveSectionActive && (
        <>
          <div className="matter-sidebar-archived-matters-selectable">
            <div className="matter-sidebar-archived-matters-selectable-section">
              <SelectableItem onSelect={toggleArchivedMattersView} dataTestId={`archived-matters-selectable`}>
                <div className="matter-sidebar-archived-matters-selectable-item">
                  <div>
                    <Icon appearance="light" size={16} iconName="box-archive" />
                    <Text appearance="body-strong-md" id="chat-archived-matters-navigation">
                      Archived Matters
                    </Text>
                  </div>

                  <Icon
                    className="matter-sidebar-archived-matters-selectable-icon"
                    appearance="light"
                    size={16}
                    iconName="chevron-right"
                  />
                </div>
              </SelectableItem>
            </div>
          </div>
        </>
      )}

      {matterSettingsActive && (
        <>
          <div className="matter-sidebar-footer-button-container">
            <SelectableItem
              onSelect={onOpenMatterFiles}
              selected={isMatterFilesActive}
              dataTestId="matter-files-button"
            >
              <Icon appearance={isMatterFilesActive ? 'solid' : 'light'} size={16} iconName={'files'} />
              <Text appearance="body-default-md" id={`matter-files-button`}>
                <Translate tKey="MATTER_FILES" />
              </Text>
            </SelectableItem>
            <SelectableItem
              onSelect={onOpenSettings}
              selected={isMatterSettingsActive}
              dataTestId="matter-settings-button"
            >
              <Icon appearance={isMatterSettingsActive ? 'solid' : 'light'} size={16} iconName={'gear'} />
              <Text appearance="body-default-md" id={`matter-settings-button`}>
                <Translate tKey="MATTER_SETTINGS" />
              </Text>
            </SelectableItem>
          </div>
        </>
      )}
    </div>
  );
};

export default MatterSidebarFooter;
