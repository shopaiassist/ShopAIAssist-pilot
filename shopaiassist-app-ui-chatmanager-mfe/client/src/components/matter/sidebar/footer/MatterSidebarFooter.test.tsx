import { act, renderWithI18n, snapshot, waitFor } from '../../../../../jest-utils';
import { FolderItem } from '../../../../@types/sidebar';
import MatterSidebarFooter from './MatterSidebarFooter';

describe('MatterSidebarFooter Tests', () => {
  const mockedChat: FolderItem = {
    treeItemId: 'fake-chat-id',
    name: 'fakeChat',
    createdAt: new Date(),
    updatedAt: new Date(),
    type: 'chat',
    fileCollectionId: 'whatever',
    matterId: 'fake-matter-id',
    description: 'fake-description',
    isArchived: false
  };

  const getMockedComponent = (
    mockOpenMatterSetting: () => void,
    mockOpenMatterFiles: () => void,
    mockToggleArchivedMattersView: () => void,
    mockedActiveFolder?: FolderItem,
    mockedArchivedMattersView: boolean = true
  ) => {
    return (
      <MatterSidebarFooter
        onOpenSettings={mockOpenMatterSetting}
        onOpenMatterFiles={mockOpenMatterFiles}
        toggleArchivedMattersView={mockToggleArchivedMattersView}
        archivedMattersView={mockedArchivedMattersView}
        activeFolder={mockedActiveFolder}
      />
    );
  };

  describe('render', () => {
    it('should render', async () => {
      snapshot(
        getMockedComponent(
          () => {},
          () => {},
          () => {}
        )
      );
    });
  });

  describe('user actions', () => {
    const mockOpenMatterSetting = jest.fn();
    const mockOpenMatterFiles = jest.fn();
    const mockToggleArchivedMattersView = jest.fn();

    it('should open call to open matter settings', async () => {
      const component = getMockedComponent(
        mockOpenMatterSetting,
        () => {},
        () => {},
        mockedChat
      );
      const wrapper = renderWithI18n(component);
      const matterSettingsButton = await wrapper.getByTestId('matter-settings-button');
      matterSettingsButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      expect(mockOpenMatterSetting).toHaveBeenCalled();
    });

    it('should open call to open matter files', async () => {
      const component = getMockedComponent(
        () => {},
        mockOpenMatterFiles,
        () => {},
        mockedChat
      );
      const wrapper = renderWithI18n(component);
      const matterFilesButton = await wrapper.getByTestId('matter-files-button');
      matterFilesButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      expect(mockOpenMatterFiles).toHaveBeenCalled();
    });

    it('should open call to archived matters', async () => {
      const component = getMockedComponent(
        () => {},
        () => {},
        mockToggleArchivedMattersView,
        undefined,
        false
      );
      const wrapper = renderWithI18n(component);
      const matterFilesButton = await wrapper.getByTestId('archived-matters-selectable');
      matterFilesButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      expect(mockToggleArchivedMattersView).toHaveBeenCalled();
    });
  });
});
