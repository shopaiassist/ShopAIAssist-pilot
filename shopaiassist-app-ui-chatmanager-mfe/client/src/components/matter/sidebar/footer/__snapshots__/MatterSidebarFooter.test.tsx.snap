// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatterSidebarFooter Tests render should render 1`] = `
<div
  class="matter-sidebar-footer"
/>
`;
